apollo-router/src/compute_job.rs

```diff
@@ -32,6 +33,31 @@ fn thread_pool_size() -> usize {
     }
 }
 
+pub(crate) struct JobStatus<'a, T> {
+    result_sender: &'a oneshot::Sender<std::thread::Result<T>>,
+}
+
+impl<T> JobStatus<'_, T> {
+    /// Checks whether the oneshot receiver for the result of the job was dropped,
+    /// which means nothing is expecting the result anymore.
+    ///
+    /// This can happen if the Tokio task owning it is cancelled,
+    /// such as if a supergraph client disconnects or if a request times out.
+    ///
+    /// In this case, a long-running job should try to cancel itself
+    /// to avoid needless resource consumption.
+    pub(crate) fn check_for_cooperative_cancellation(&self) -> ControlFlow<()> {
```

Comment:
Pretty cool! While tying it to the client request (disconnect/time out) is a great start - if we were to expose this feature as a user specified config value it sounds like it should be as simple as wrapping those QP tasks with a [timeout](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html)?

Reply:
I don't think we should be exposing this specifically as a feature. Cancellation at the query planning level should work in the same way as backpressure in other parts of the router. That is, query planning should adhere to the router's existing `traffic_shaping.router.timeout` configuration.

Reply:
I agree that for now this feature should be tied to the request lifecycle. That being said, sometime in the future, I think we should revisit whether it would make sense to further expand on this functionality as there are some potential additional usecases (eg folks might want to still plan the query even though their request timed out with the assumption that be subsequent request would hit the cached plan, similarly we might want to have some limits when doing qp cache prewarming etc).